presto-jdbc/src/main/java/com/facebook/presto/jdbc/QueryInterceptor.java

```diff
@@ -0,0 +1,63 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.jdbc;
+
+import java.sql.Statement;
+import java.util.Map;
+
+public interface QueryInterceptor
+{
+    /**
+     * Called once per PrestoConnection at instantiation.
+     *
+     * @param properties the session properties supplied in the URI
+     * @return QueryInterceptor
+     */
+    QueryInterceptor init(Map<String, String> properties);
+
+    /**
+     * Called once to release any resources when PrestoConnection is closing.
+     */
+    void destroy();
+
+    /**
+     * Called before the query has been sent to the server. This method is only called on
+     * top-level queries i.e. queries executed by QueryInterceptors are not intercepted.
+     *
+     * This method can optionally return a PrestoResultSet. If a PrestoResultSet is returned from
+     * from preProcess then the intercepted query is not executed and that PrestoResultSet is
+     * returned instead. If there are multiple QueryInterceptors, preProcess will be called
+     * on each and the last PrestoResultSet is returned.
+     *
+     * @param sql the SQL string of the query.
+     * @param interceptedStatement the Statement being executed.
+     * @return optional PrestoResultSet to be returned
+     */
+    PrestoResultSet preProcess(String sql, Statement interceptedStatement);
+
+    /**
+     * Called after the query has been sent to the server. This method is only called on
+     * top-level queries i.e. queries executed by QueryInterceptors are not intercepted.
+     *
+     * This method can optionally return a PrestoResultSet. If a PrestoResultSet is returned
+     * from postProcess then that PrestoResultSet is returned instead. If there are multiple
+     * QueryInterceptors, postProcess will be called on each and the last PrestoResultSet is returned.
+     *
+     * @param sql the SQL string of the query.
+     * @param interceptedStatement the Statement being executed.
+     * @param originalResultSet the intercepted ResultSet
+     * @return optional PrestoResultSet to be returned instead of the original PrestoResultSet
```

Comment:
It's the same approach. I think the implementation in the BaseQueryInterceptor threw you off. In the default implementation in the interface I changed it to just return null.

Reply:
@robbypete - This code was the part that made me write this comment - 

```
if (shouldIntercept) {
                    resultSet = connection().invokeQueryInterceptorsPost(sql, this, resultSet);
                    verifyNotNull(resultSet, "invokeQueryInterceptorsPost should never return a null ResultSet");
                }
```

As I read it again carefully, it should be fine since the `processInterceptors` will just return the original ResultSet here.